File: backend/app/utils/ranges.py
from __future__ import annotations

from dataclasses import dataclass

from app.utils.timecode import parse_timecode


@dataclass(frozen=True)
class RangeSeconds:
    start: float
    end: float


def normalize_remove_ranges(
    raw_ranges: list[dict[str, str]], duration: float
) -> list[RangeSeconds]:
    parsed: list[RangeSeconds] = []
    for item in raw_ranges:
        start = parse_timecode(item["start"])
        end = parse_timecode(item["end"])
        if end <= start:
            raise ValueError(f"Range end must be greater than start ({item})")
        if start >= duration:
            continue
        parsed.append(RangeSeconds(start=max(0.0, start), end=min(duration, end)))

    if not parsed:
        return []

    parsed.sort(key=lambda r: r.start)
    merged: list[RangeSeconds] = [parsed[0]]
    for current in parsed[1:]:
        last = merged[-1]
        if current.start <= last.end:
            merged[-1] = RangeSeconds(last.start, max(last.end, current.end))
        else:
            merged.append(current)
    return merged


def complement_ranges(
    remove_ranges: list[RangeSeconds], duration: float
) -> list[RangeSeconds]:
    if duration <= 0:
        return []
    if not remove_ranges:
        return [RangeSeconds(0.0, duration)]

    keep: list[RangeSeconds] = []
    cursor = 0.0
    for removed in remove_ranges:
        if removed.start > cursor:
            keep.append(RangeSeconds(cursor, removed.start))
        cursor = max(cursor, removed.end)
    if cursor < duration:
        keep.append(RangeSeconds(cursor, duration))
    return [segment for segment in keep if segment.end - segment.start >= 0.02]

File: src/services/ranges.ts
export interface RangeSeconds {
  start: number
  end: number
}

export function normalizeRemoveRanges(
  rawRanges: Array<{ start: number; end: number }>,
  duration: number
): RangeSeconds[] {
  const parsed: RangeSeconds[] = []
  for (const item of rawRanges) {
    if (item.end <= item.start) {
      throw new Error(
        `Range end must be greater than start (${item.start} - ${item.end})`
      )
    }
    if (item.start >= duration) continue
    parsed.push({
      start: Math.max(0, item.start),
      end: Math.min(duration, item.end),
    })
  }

  if (parsed.length === 0) return []

  parsed.sort((a, b) => a.start - b.start)
  const merged: RangeSeconds[] = [parsed[0]]
  for (let i = 1; i < parsed.length; i++) {
    const current = parsed[i]
    const last = merged[merged.length - 1]
    if (current.start <= last.end) {
      last.end = Math.max(last.end, current.end)
    } else {
      merged.push(current)
    }
  }
  return merged
}

export function complementRanges(
  removeRanges: RangeSeconds[],
  duration: number
): RangeSeconds[] {
  if (duration <= 0) return []
  if (removeRanges.length === 0) return [{ start: 0, end: duration }]

  const keep: RangeSeconds[] = []
  let cursor = 0
  for (const removed of removeRanges) {
    if (removed.start > cursor) {
      keep.push({ start: cursor, end: removed.start })
    }
    cursor = Math.max(cursor, removed.end)
  }
  if (cursor < duration) {
    keep.push({ start: cursor, end: duration })
  }
  return keep.filter((seg) => seg.end - seg.start >= 0.02)
}
